File: backend/design_intelligence.py
"""QRU Design Intelligence™ — the permanent visual foundation of Creative Studio™.

Creative Studio does not copy previous designs; it LEARNS the QRU Design Language™ and
consistently produces original products that feel unmistakably like QRU.

Libraries (Mongo collections):
- brand_library    : the canonical QRU brand standards (colors, typography, shield, badge, spacing, components, templates)
- design_library   : approved design references — grows as products earn Treasure Standard™
- master_assets    : the Master Asset Library™ (searchable, reusable assets with full metadata)
- design_language  : learned, evolving design principles
"""
import logging
from database import db
from models import gen_id, now_iso, clean

logger = logging.getLogger("qru.design")

SHIELD = "/qru-shield-light.png"

# Canonical QRU brand standards (the source of truth for every product).
BRAND_STANDARDS = {
    "colors": [
        {"name": "Royal Purple", "role": "Primary", "hsl": "266 74% 24%", "hex": "#35106A"},
        {"name": "QRU Gold", "role": "Accent", "hsl": "42 91% 53%", "hex": "#F5B21A"},
        {"name": "Deep Navy", "role": "Depth", "hsl": "252 45% 18%", "hex": "#221A42"},
        {"name": "White", "role": "Space", "hsl": "0 0% 100%", "hex": "#FFFFFF"},
        {"name": "Success Green", "role": "Verified", "hsl": "160 84% 34%", "hex": "#0E9F6E"},
    ],
    "typography": [
        {"name": "Outfit", "role": "Headings", "weights": [600, 700, 800]},
        {"name": "IBM Plex Sans", "role": "Body", "weights": [400, 500, 600]},
    ],
    "spacing": {"principle": "Generous — 2-3x breathing room reduces cognitive load", "unit": 4, "scale": [4, 8, 12, 16, 24, 32, 48, 64]},
    "shield": {"name": "QRU Shield™", "asset": SHIELD, "placement": "Top-left of every product"},
    "treasure_badge": {"name": "Treasure Standard™ Badge", "placement": "Displayed only on certified products"},
    "principles": [
        "Unmistakably QRU — brand recognizable at a glance",
        "Clarity over decoration — design supports understanding",
        "Warm, premium, trustworthy, hopeful",
        "Left-aligned reading flow with clear information hierarchy",
        "Memory Sentence™ always in a Royal Purple callout",
        "Verified status always visible; references expand on request",
    ],
    "components": [
        {"name": "QRU Callout Box", "use": "Memory Sentence™ and key takeaways"},
        {"name": "Verification Card", "use": "Kingdom Lion™ status + evidence level"},
        {"name": "Layer Selector", "use": "Layered Understanding depth control"},
        {"name": "Treasure Ribbon", "use": "Treasure Standard™ certification marker"},
        {"name": "QRU Product Card", "use": "Catalog and related products"},
    ],
    "templates": [
        {"name": "Master Cover Template™", "for": ["Book", "Course", "Workbook"]},
        {"name": "Poster Template™", "for": ["Poster", "Infographic"]},
        {"name": "Workbook Layout™", "for": ["Workbook", "Lesson Plan"]},
        {"name": "Presentation Layout™", "for": ["Presentation"]},
        {"name": "Interactive Lesson Template™", "for": ["Interactive Lesson", "Quiz", "Flash Cards"]},
        {"name": "Guide Template™", "for": ["Teacher Guide", "Caregiver Guide"]},
    ],
}

# Creative Studio autonomy — automatic template + style selection by product type.
TEMPLATE_RULES = {
    "Book": {"template": "Master Cover Template™", "layout": "Chaptered", "illustration": "Editorial"},
    "Workbook": {"template": "Workbook Layout™", "layout": "Activity Grid", "illustration": "Friendly"},
    "Poster": {"template": "Poster Template™", "layout": "Single Focal", "illustration": "Bold"},
    "Infographic": {"template": "Poster Template™", "layout": "Flow Diagram", "illustration": "Iconographic"},
    "Presentation": {"template": "Presentation Layout™", "layout": "Slide Deck", "illustration": "Minimal"},
    "Teacher Guide": {"template": "Guide Template™", "layout": "Structured", "illustration": "Editorial"},
    "Caregiver Guide": {"template": "Guide Template™", "layout": "Conversational", "illustration": "Warm"},
    "Course": {"template": "Master Cover Template™", "layout": "Modular", "illustration": "Editorial"},
    "Interactive Lesson": {"template": "Interactive Lesson Template™", "layout": "Progressive", "illustration": "Friendly"},
    "Quiz": {"template": "Interactive Lesson Template™", "layout": "Question Cards", "illustration": "Playful"},
    "Flash Cards": {"template": "Interactive Lesson Template™", "layout": "Card Deck", "illustration": "Playful"},
    "Podcast Script": {"template": "Guide Template™", "layout": "Script", "illustration": "Minimal"},
    "Video Script": {"template": "Guide Template™", "layout": "Storyboard", "illustration": "Cinematic"},
    "Short-form Content": {"template": "Poster Template™", "layout": "Social Tile", "illustration": "Bold"},
}

AUDIENCE_STYLE = {
    "Children": {"palette": "Bright QRU (Gold-forward)", "typography": "Outfit rounded", "tone": "Playful & encouraging"},
    "Teens": {"palette": "Modern QRU (Purple-forward)", "typography": "Outfit bold", "tone": "Confident & energetic"},
    "General public": {"palette": "Core QRU", "typography": "Outfit + IBM Plex", "tone": "Warm & clear"},
    "Professionals": {"palette": "Navy-forward QRU", "typography": "IBM Plex refined", "tone": "Precise & trusted"},
    "Caregivers": {"palette": "Warm QRU", "typography": "IBM Plex", "tone": "Reassuring"},
    "Teachers": {"palette": "Core QRU", "typography": "Outfit + IBM Plex", "tone": "Structured & supportive"},
}

# The QRU Design Checklist™ — must all pass before a design is approved.
CHECKLIST_QUESTIONS = [
    "Is this unmistakably QRU?",
    "Would someone recognize the brand immediately?",
    "Does it follow QRU typography?",
    "Does it follow QRU colors?",
    "Does it support understanding?",
    "Does it reduce confusion?",
    "Does it feel premium?",
    "Does it create emotional trust?",
    "Does it reinforce the QRU mission?",
    "Would someone proudly share this product?",
]

BASELINE_PRINCIPLES = [
    {"principle": "QRU Shield™ anchors the top-left of every product", "source": "Brand Standard", "category": "Branding"},
    {"principle": "Memory Sentence™ lives in a Royal Purple callout", "source": "Brand Standard", "category": "Layout"},
    {"principle": "Headings in Outfit, body in IBM Plex Sans", "source": "Brand Standard", "category": "Typography"},
    {"principle": "Generous spacing reduces cognitive load", "source": "Brand Standard", "category": "Spacing"},
    {"principle": "Verified status is always visible; references expand on request", "source": "Brand Standard", "category": "Trust"},
]


async def seed_design_intelligence():
    if await db.brand_library.count_documents({}) == 0:
        await db.brand_library.insert_one({"id": gen_id(), "kind": "standards", **BRAND_STANDARDS, "created_at": now_iso()})
    if await db.design_language.count_documents({}) == 0:
        for p in BASELINE_PRINCIPLES:
            await db.design_language.insert_one({"id": gen_id(), **p, "created_at": now_iso()})
    # seed master assets with the core brand assets
    if await db.master_assets.count_documents({}) == 0:
        core = [
            ("QRU Shield™", "Logo", "QRU Brand", ["shield", "logo", "brand"], SHIELD),
            ("Treasure Standard™ Badge", "Badge", "QRU Brand", ["treasure", "badge", "quality"], SHIELD),
            ("Master Cover Template™", "Template", "QRU Brand", ["cover", "template", "book"], SHIELD),
        ]
        for title, cat, fam, kw, preview in core:
            await db.master_assets.insert_one({
                "id": gen_id(), "asset_id": f"QRU-ASSET-{gen_id()[:6].upper()}", "title": title,
                "category": cat, "product_family": fam, "knowledge_record_id": None, "keywords": kw,
                "version": 1, "treasure_standard": True, "creator": "Creative Studio Director™",
                "related_products": [], "preview": preview, "editable_source": None, "export_files": [],
                "license_status": "QRU Owned", "created_at": now_iso(),
            })
    # learn from any existing Treasure Standard™ products so the library isn't empty
    if await db.design_library.count_documents({}) == 0:
        approved = await db.products.find({"treasure_standard": True, "status": "Published"}).to_list(50)
        for p in approved:
            await learn_from_product(p)


def recommend_templates(product_type, audience="General public"):
    rule = TEMPLATE_RULES.get(product_type, {"template": "Master Cover Template™", "layout": "Standard", "illustration": "Editorial"})
    style = AUDIENCE_STYLE.get(audience, AUDIENCE_STYLE["General public"])
    return {
        "product_type": product_type, "audience": audience,
        "cover_template": rule["template"], "layout": rule["layout"], "illustration_style": rule["illustration"],
        "palette": style["palette"], "typography": style["typography"], "tone": style["tone"],
        "icon_set": "QRU Line Icons", "brand_elements": ["QRU Shield™", "Brand Footer", "Treasure Standard™ Badge"],
    }


async def run_design_checklist(product):
    """Automated Design Checklist™. For QRU-manufactured products these are deterministic,
    driven by whether Creative Studio branding + Treasure Standard™ are applied."""
    branded = product.get("creative_status") == "Reviewed"
    treasure = product.get("treasure_standard", False)
    traceable = bool(product.get("knowledge_record_id"))
    answers = []
    for q in CHECKLIST_QUESTIONS:
        # Understanding/mission questions rely on traceable verified knowledge; visual questions on branding.
        if q in ("Does it support understanding?", "Does it reduce confusion?", "Does it reinforce the QRU mission?"):
            passed = traceable
        elif q in ("Does it feel premium?", "Would someone proudly share this product?"):
            passed = branded and treasure
        else:
            passed = branded
        answers.append({"question": q, "pass": passed})
    all_pass = all(a["pass"] for a in answers)
    return {"passed": all_pass, "answers": answers,
            "verdict": "Approved — unmistakably QRU." if all_pass else "Revise — apply QRU branding before release."}


async def learn_from_product(product):
    """Called when a product earns Treasure Standard™. Grows the Design Library, registers a
    Master Asset, and extracts a reusable design principle — the system learns from itself."""
    try:
        pid = product["id"]
        if await db.design_library.find_one({"product_id": pid}):
            return
        rule = TEMPLATE_RULES.get(product.get("product_type"), {})
        # Design Library reference
        await db.design_library.insert_one({
            "id": gen_id(), "product_id": pid, "product_code": product.get("product_code"),
            "title": product.get("title"), "product_type": product.get("product_type"),
            "product_family": product.get("family"), "knowledge_record_id": product.get("knowledge_record_id"),
            "template": rule.get("template"), "layout": rule.get("layout"),
            "treasure_standard": True, "preview": SHIELD, "created_at": now_iso(),
        })
        # Master Asset entry
        await db.master_assets.insert_one({
            "id": gen_id(), "asset_id": f"QRU-ASSET-{gen_id()[:6].upper()}",
            "title": product.get("title"), "category": product.get("product_type"),
            "product_family": product.get("family"), "knowledge_record_id": product.get("knowledge_record_id"),
            "keywords": [product.get("product_type", ""), product.get("family", ""), "treasure standard"],
            "version": 1, "treasure_standard": True, "creator": "Creative Studio Director™",
            "related_products": [], "preview": SHIELD, "editable_source": None, "export_files": [],
            "license_status": "QRU Owned", "created_at": now_iso(),
        })
        # Extract a design principle (learning) — bounded, no duplicate spam
        ptype = product.get("product_type")
        existing = await db.design_language.find_one({"category": "Learned", "principle": {"$regex": ptype}})
        if not existing and rule.get("template"):
            await db.design_language.insert_one({
                "id": gen_id(),
                "principle": f"{ptype} products use the {rule.get('template')} with a {rule.get('layout','standard')} layout for QRU consistency.",
                "source": product.get("product_code", "Approved product"), "category": "Learned", "created_at": now_iso(),
            })
    except Exception as e:
        logger.error(f"design learning failed: {e}")


async def library_stats():
    return {
        "brand_standards": 1,
        "design_references": await db.design_library.count_documents({}),
        "master_assets": await db.master_assets.count_documents({}),
        "design_principles": await db.design_language.count_documents({}),
    }

File: frontend/src/pages/DesignIntelligence.js
import { useEffect, useState } from "react";
import { Loader2, Palette, Library, Layers, Search, Sparkles, CheckCircle2, Type, Grid3x3 } from "lucide-react";
import api from "@/lib/api";
import { PageHeader } from "@/components/shared";

const TABS = [
  { key: "brand", label: "Brand Library™", icon: Palette },
  { key: "design", label: "Design Library™", icon: Library },
  { key: "assets", label: "Master Assets™", icon: Grid3x3 },
  { key: "language", label: "Design Language™", icon: Type },
  { key: "autonomy", label: "Studio Autonomy", icon: Sparkles },
];

export default function DesignIntelligence() {
  const [tab, setTab] = useState("brand");
  const [stats, setStats] = useState(null);
  const [brand, setBrand] = useState(null);
  const [refs, setRefs] = useState([]);
  const [assets, setAssets] = useState([]);
  const [principles, setPrinciples] = useState([]);
  const [q, setQ] = useState("");
  const [rec, setRec] = useState(null);
  const [recType, setRecType] = useState("Poster");
  const [recAud, setRecAud] = useState("Children");

  useEffect(() => {
    api.get("/design/stats").then((r) => setStats(r.data));
    api.get("/design/brand-library").then((r) => setBrand(r.data));
    api.get("/design/design-library").then((r) => setRefs(r.data.references));
    api.get("/design/master-assets").then((r) => setAssets(r.data.assets));
    api.get("/design/design-language").then((r) => setPrinciples(r.data.principles));
  }, []);

  const searchAssets = async (val) => {
    setQ(val);
    const { data } = await api.get("/design/master-assets", { params: val ? { q: val } : {} });
    setAssets(data.assets);
  };

  const recommend = async () => {
    const { data } = await api.get("/design/recommend-templates", { params: { product_type: recType, audience: recAud } });
    setRec(data);
  };

  if (!brand) return <div className="flex justify-center py-32"><Loader2 className="w-6 h-6 animate-spin text-primary" /></div>;

  return (
    <div>
      <PageHeader overline="Creative Studio™ · Enterprise Mode" title="QRU Design Intelligence™"
        description="Creative Studio learns the QRU Design Language™ so every future product feels unmistakably QRU — original, never copied." />

      {stats && (
        <div className="grid grid-cols-2 sm:grid-cols-4 gap-3 mb-6">
          {[["Design References", stats.design_references], ["Master Assets", stats.master_assets], ["Design Principles", stats.design_principles], ["Brand Standards", stats.brand_standards]].map(([l, v]) => (
            <div key={l} className="bg-card border rounded-xl p-4"><p className="font-heading text-2xl font-bold">{v}</p><p className="text-[11px] text-muted-foreground">{l}</p></div>
          ))}
        </div>
      )}

      <div className="flex items-center gap-1.5 flex-wrap mb-6 border-b border-border">
        {TABS.map((t) => (
          <button key={t.key} data-testid={`di-tab-${t.key}`} onClick={() => setTab(t.key)}
            className={`flex items-center gap-1.5 px-3 py-2 text-sm font-medium border-b-2 -mb-px transition-colors ${tab === t.key ? "border-primary text-primary" : "border-transparent text-muted-foreground hover:text-foreground"}`}>
            <t.icon className="w-4 h-4" />{t.label}
          </button>
        ))}
      </div>

      {tab === "brand" && (
        <div className="space-y-8" data-testid="di-brand">
          <div>
            <p className="overline text-primary mb-3">Color System</p>
            <div className="grid grid-cols-2 sm:grid-cols-5 gap-3">
              {brand.colors?.map((c) => (
                <div key={c.name} className="rounded-xl overflow-hidden border">
                  <div className="h-16" style={{ background: `hsl(${c.hsl})` }} />
                  <div className="p-2"><p className="text-xs font-semibold">{c.name}</p><p className="text-[10px] text-muted-foreground">{c.role} · {c.hex}</p></div>
                </div>
              ))}
            </div>
          </div>
          <div className="grid sm:grid-cols-2 gap-6">
            <div>
              <p className="overline text-primary mb-3">Typography</p>
              {brand.typography?.map((t) => (
                <div key={t.name} className="bg-card border rounded-xl p-4 mb-2">
                  <p className="font-heading text-lg font-bold">{t.name}</p>
                  <p className="text-xs text-muted-foreground">{t.role} · weights {t.weights?.join(", ")}</p>
                </div>
              ))}
            </div>
            <div>
              <p className="overline text-primary mb-3">Design Principles</p>
              <ul className="space-y-1.5 text-sm">{brand.principles?.map((p, i) => <li key={i} className="flex gap-2"><CheckCircle2 className="w-4 h-4 shrink-0" style={{ color: "hsl(var(--success))" }} />{p}</li>)}</ul>
            </div>
          </div>
          <div className="grid sm:grid-cols-2 gap-6">
            <div>
              <p className="overline text-primary mb-3">Components</p>
              <div className="space-y-1.5">{brand.components?.map((c) => <div key={c.name} className="bg-card border rounded-lg p-3"><p className="text-sm font-semibold">{c.name}</p><p className="text-xs text-muted-foreground">{c.use}</p></div>)}</div>
            </div>
            <div>
              <p className="overline text-primary mb-3">Templates</p>
              <div className="space-y-1.5">{brand.templates?.map((t) => <div key={t.name} className="bg-card border rounded-lg p-3"><p className="text-sm font-semibold">{t.name}</p><p className="text-xs text-muted-foreground">For: {t.for?.join(", ")}</p></div>)}</div>
            </div>
          </div>
        </div>
      )}

      {tab === "design" && (
        <div data-testid="di-design">
          <p className="text-sm text-muted-foreground mb-4">Every Treasure Standard™ product becomes a reusable design reference. The more QRU creates, the smarter Creative Studio becomes.</p>
          {refs.length === 0 ? <p className="text-muted-foreground py-12 text-center">No references yet — certify products to grow the library.</p> : (
            <div className="grid gap-4 sm:grid-cols-2 lg:grid-cols-3">
              {refs.map((r) => (
                <div key={r.id} className="bg-card border rounded-xl overflow-hidden" data-testid={`di-ref-${r.id}`}>
                  <div className="h-24 flex items-center justify-center" style={{ background: "hsl(var(--secondary))" }}><img src={r.preview} alt="" className="w-10 h-10 object-contain" /></div>
                  <div className="p-3">
                    <p className="text-sm font-semibold line-clamp-1">{r.title}</p>
                    <p className="text-[11px] text-muted-foreground">{r.product_type} · {r.template || "QRU Template"}</p>
                  </div>
                </div>
              ))}
            </div>
          )}
        </div>
      )}

      {tab === "assets" && (
        <div data-testid="di-assets">
          <div className="relative max-w-md mb-4">
            <Search className="w-4 h-4 absolute left-3 top-1/2 -translate-y-1/2 text-muted-foreground" />
            <input data-testid="di-asset-search" value={q} onChange={(e) => searchAssets(e.target.value)} placeholder="Search assets, keywords, IDs…"
              className="w-full pl-9 pr-3 py-2 text-sm bg-muted rounded-md border border-transparent focus:border-primary outline-none" />
          </div>
          <div className="grid gap-3 sm:grid-cols-2 lg:grid-cols-3">
            {assets.map((a) => (
              <div key={a.id} className="bg-card border rounded-xl p-3 flex gap-3" data-testid={`di-asset-${a.id}`}>
                <img src={a.preview} alt="" className="w-10 h-10 object-contain shrink-0" />
                <div className="min-w-0">
                  <p className="text-sm font-semibold line-clamp-1">{a.title}</p>
                  <p className="text-[10px] text-muted-foreground">{a.asset_id}</p>
                  <p className="text-[11px] text-muted-foreground">{a.category} · {a.product_family} · {a.license_status}</p>
                </div>
              </div>
            ))}
          </div>
        </div>
      )}

      {tab === "language" && (
        <div data-testid="di-language">
          <p className="text-sm text-muted-foreground mb-4">The evolving QRU Design Language™ — baseline standards plus principles learned from approved products.</p>
          <div className="space-y-2">
            {principles.map((p) => (
              <div key={p.id} className="bg-card border rounded-xl p-4 flex items-start gap-3">
                <span className={`text-[10px] px-2 py-0.5 rounded-full shrink-0 mt-0.5 ${p.category === "Learned" ? "bg-gold/15" : "bg-secondary text-primary"}`} style={p.category === "Learned" ? { color: "hsl(var(--navy))" } : {}}>{p.category}</span>
                <div><p className="text-sm">{p.principle}</p><p className="text-[11px] text-muted-foreground mt-0.5">Source: {p.source}</p></div>
              </div>
            ))}
          </div>
        </div>
      )}

      {tab === "autonomy" && (
        <div data-testid="di-autonomy">
          <p className="text-sm text-muted-foreground mb-4">Creative Studio automatically selects the right template, palette, typography, and illustration style — no manual selection needed.</p>
          <div className="bg-card border rounded-2xl p-5 max-w-2xl">
            <div className="grid sm:grid-cols-2 gap-3 mb-4">
              <div>
                <label className="text-xs text-muted-foreground mb-1 block">Product Type</label>
                <select data-testid="di-rec-type" value={recType} onChange={(e) => setRecType(e.target.value)} className="w-full px-3 py-2 text-sm bg-muted rounded-md border outline-none">
                  {["Book", "Workbook", "Poster", "Infographic", "Presentation", "Teacher Guide", "Caregiver Guide", "Course", "Interactive Lesson", "Quiz", "Flash Cards", "Video Script"].map((t) => <option key={t}>{t}</option>)}
                </select>
              </div>
              <div>
                <label className="text-xs text-muted-foreground mb-1 block">Audience</label>
                <select data-testid="di-rec-aud" value={recAud} onChange={(e) => setRecAud(e.target.value)} className="w-full px-3 py-2 text-sm bg-muted rounded-md border outline-none">
                  {["Children", "Teens", "General public", "Professionals", "Caregivers", "Teachers"].map((a) => <option key={a}>{a}</option>)}
                </select>
              </div>
            </div>
            <button data-testid="di-recommend" onClick={recommend} className="px-5 py-2 rounded-md text-white text-sm font-medium" style={{ background: "hsl(var(--royal))" }}>Auto-Select Design</button>
            {rec && (
              <div className="mt-5 grid sm:grid-cols-2 gap-3 text-sm" data-testid="di-rec-result">
                {[["Cover Template", rec.cover_template], ["Layout", rec.layout], ["Illustration", rec.illustration_style], ["Palette", rec.palette], ["Typography", rec.typography], ["Tone", rec.tone], ["Icon Set", rec.icon_set], ["Brand Elements", rec.brand_elements?.join(", ")]].map(([l, v]) => (
                  <div key={l} className="border rounded-lg p-3"><p className="text-[11px] text-muted-foreground">{l}</p><p className="font-medium">{v}</p></div>
                ))}
              </div>
            )}
          </div>
        </div>
      )}
    </div>
  );
}
